src/hotspot/share/classfile/systemDictionaryShared.cpp

```diff
@@ -1053,11 +1053,17 @@ InstanceKlass* SystemDictionaryShared::find_or_load_shared_class(
   return k;
 }
 
-PackageEntry* SystemDictionaryShared::get_package_entry_from_class_name(Handle class_loader, Symbol* class_name) {
-  PackageEntry* pkg_entry = NULL;
-  TempNewSymbol pkg_name = ClassLoader::package_from_class_name(class_name);
+PackageEntry* SystemDictionaryShared::get_package_entry_from_class_name(InstanceKlass* ik, Handle class_loader) {
+  PackageEntry* pkg_entry = ik->package();
+  if (MetaspaceShared::use_full_module_graph() && ik->is_shared() &&
+       ((pkg_entry == NULL) || MetaspaceShared::is_in_shared_metaspace(pkg_entry))) {
```

Comment:
I think the `(pkg_entry == NULL)` check will be incorrect if the class is loaded by a custom loader, or is in an unnamed module.

Also, I wonder if you added the check `MetaspaceShared::is_in_shared_metaspace(pkg_entry)` because of the missing `ArchiveUtils::mark_pointer()` call.

How about changing to

```
if (MetaspaceShared::use_full_module_graph() && ik->is_shared() && pkg_entry != NULL) {
  assert(MetaspaceShared::is_in_shared_metaspace(pkg_entry), "must be");
  // also assert not in unnamed module, and not unregistered loader ...
  return pkg_entry;
}
```